README.md
# edtf

A library for parsing dates using the [EDTF](https://www.loc.gov/standards/datetime/edtf.html) standard.

## Usage

A simple usage example:

```dart
import 'package:edtf/edtf.dart';

main() {
  /// Parse a date in the EDTF format
  final date = EdtfDate.parse('1985-05%-13?');

  /// Check that the year is an approximation.
  if (date.year.approx) {
    /// If it is an approximation, then print it
    print(date.toString());
  }

  /// Parse an EDTF formatted interval
  final interval = EdtfInterval.parse('../2012');

  /// If the interval is open started and the end is a known date...
  if (interval.openStart || interval.end != null) {
    /// ... print the end date.
    print(interval.end.toString());
  }

  /// Parse a set of EDTF dates/intervals, where the real date is one of them.
  final oneOf = EdtfOneOf.parse('[1981, 1982-~04, 1983-1987]');

  /// Iterate through the dates/intervals
  for (final elem in oneOf.values) {
    /// If the elem is a date print it in this format
    if (elem is EdtfDate) {
      print('Date: ' + elem.toString());
    }

    /// If the elem is an interval just print the start date
    if (elem is EdtfInterval) {
      print('Interval start date: ' +
          (elem.openStart ? 'open' : elem.start.toString()));
      // An interval either has an known start date or open started
    }
  }
}
```

CHANGELOG.md
## 1.1.0

- Default constructors are hidden.

## 1.0.0

- EDTF parsing and toString working


test/edtf_test.dart
import 'package:edtf/edtf.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
//    setUp(() {
//    });
    test('Date parse test', () {
      ///                      Y-12019-%05-26~T15:34:06Z
      final dateParse = Edtf.parse('Y-12019-?05-26~T15:34:06Z');
      expect(dateParse.runtimeType, EdtfDate);
      final date = dateParse as EdtfDate;
      expect(date.year.value, -12019);
      expect(date.year.localApprox, false);
      expect(date.year.localUncert, false);
      expect(date.year.groupApprox, true);
      expect(date.year.groupUncert, false);
      expect(date.month.value, 5);
      expect(date.month.localApprox, false);
      expect(date.month.localUncert, true);
      expect(date.month.groupApprox, true);
      expect(date.month.groupUncert, false);
      expect(date.day.value, 26);
      expect(date.day.localApprox, true);
      expect(date.day.localUncert, false);
      expect(date.day.groupApprox, true);
      expect(date.day.groupUncert, false);
      expect(date.time.hour, 15);
      expect(date.time.minutes, 34);
      expect(date.time.seconds, 6);
      expect(date.time.shiftLevel, EdtfTime.shiftLevelUTC);
    });
    test('Date year parse test', () {
      final year1 = Edtf.parse('1234') as EdtfDate;
      assert(year1.year != null);
      assert(year1.month == null);
      assert(year1.day == null);
      assert(year1.time == null);
      expect(year1.year.value, 1234);
      expect(year1.year.exp, null);
      expect(year1.year.significance, null);
      expect(year1.year.unspecMask, '1234');
      expect(year1.year.localApprox, false);
      expect(year1.year.localUncert, false);
      expect(year1.year.groupApprox, false);
      expect(year1.year.groupUncert, false);
      final year2 = Edtf.parse('~-20XX?') as EdtfDate;
      assert(year2.year != null);
      assert(year2.month == null);
      assert(year2.day == null);
      assert(year2.time == null);
      expect(year2.year.value, -2000);
      expect(year2.year.exp, null);
      expect(year2.year.significance, null);
      expect(year2.year.unspecMask, '-20..');
      expect(year2.year.localApprox, true);
      expect(year2.year.localUncert, true);
      expect(year2.year.groupApprox, false);
      expect(year2.year.groupUncert, true);
      final year3 = Edtf.parse('-12X3E9S3%') as EdtfDate;
      assert(year3.year != null);
      assert(year3.month == null);
      assert(year3.day == null);
      assert(year3.time == null);
      expect(year3.year.value, -1203);
      expect(year3.year.exp, 9);
      expect(year3.year.significance, 3);
      expect(year3.year.unspecMask, '-12.3');
      expect(year3.year.localApprox, true);
      expect(year3.year.localUncert, true);
      expect(year3.year.groupApprox, true);
      expect(year3.year.groupUncert, true);
    });
    test('Interval parse test', () {
      final uninterval = Edtf.parse('/') as EdtfInterval;
      expect(uninterval.start, null);
      expect(uninterval.openStart, false);
      expect(uninterval.end, null);
      expect(uninterval.openEnd, false);
      final openinterval = Edtf.parse('../..') as EdtfInterval;
      expect(openinterval.start, null);
      expect(openinterval.openStart, true);
      expect(openinterval.end, null);
      expect(openinterval.openEnd, true);
      final ival = Edtf.parse('1900/2019') as EdtfInterval;
      expect(ival.start.year.value, 1900);
      expect(ival.openStart, false);
      expect(ival.end.year.value, 2019);
      expect(ival.openEnd, false);
    });
    test('Set parse tests', () {
      final oneOf = Edtf.parse('[1999, 2000, 2013..2017, 2019..]') as EdtfOneOf;
      expect(oneOf.values[0].runtimeType, EdtfDate);
      expect(oneOf.values[1].runtimeType, EdtfDate);
      expect(oneOf.values[2].runtimeType, EdtfInterval);
      expect(oneOf.values[3].runtimeType, EdtfInterval);
      expect((oneOf.values[0] as EdtfDate).year.value, 1999);
      expect((oneOf.values[1] as EdtfDate).year.value, 2000);
      expect((oneOf.values[2] as EdtfInterval).start.year.value, 2013);
      expect((oneOf.values[2] as EdtfInterval).end.year.value, 2017);
      expect((oneOf.values[3] as EdtfInterval).start.year.value, 2019);
      expect((oneOf.values[3] as EdtfInterval).openEnd, true);
    });

    test('Back-to-string tests', () {
      final e1_1 = Edtf.parse('1985-04-12');
      expect(e1_1.toString(), '1985-04-12');
      final e1_2 = Edtf.parse('1985-04');
      expect(e1_2.toString(), '1985-04');
      final e1_3 = Edtf.parse('1985');
      expect(e1_3.toString(), '1985');
      final e2_1 = Edtf.parse('1985-04-12T23:20:30');
      expect(e2_1.toString(), '1985-04-12T23:20:30');
      final e2_2 = Edtf.parse('1985-04-12T23:20:30Z');
      expect(e2_2.toString(), '1985-04-12T23:20:30Z');
      final e2_3 = Edtf.parse('1985-04-12T23:20:30-04');
      expect(e2_3.toString(), '1985-04-12T23:20:30-04');
      final e2_4 = Edtf.parse('1985-04-12T23:20:30+04:30');
      expect(e2_4.toString(), '1985-04-12T23:20:30+04:30');
      final e3_1 = Edtf.parse('1964/2008');
      expect(e3_1.toString(), '1964/2008');
      final e3_2 = Edtf.parse('2004-06/2006-08');
      expect(e3_2.toString(), '2004-06/2006-08');
      final e3_3 = Edtf.parse('2004-02-01/2005-02-08');
      expect(e3_3.toString(), '2004-02-01/2005-02-08');
      final e3_4 = Edtf.parse('2004-02-01/2005-02');
      expect(e3_4.toString(), '2004-02-01/2005-02');
      final e3_5 = Edtf.parse('2004-02-01/2005');
      expect(e3_5.toString(), '2004-02-01/2005');
      final e3_6 = Edtf.parse('2005/2006-02');
      expect(e3_6.toString(), '2005/2006-02');
      final e4_1 = Edtf.parse('Y170000002');
      expect(e4_1.toString(), 'Y170000002');
      final e4_2 = Edtf.parse('Y-170000002');
      expect(e4_2.toString(), 'Y-170000002');
      final e5_1 = Edtf.parse('2001-21');
      expect(e5_1.toString(), '2001-21');
      final e6_1 = Edtf.parse('1984?');
      expect(e6_1.toString(), '1984?');
      final e6_2 = Edtf.parse('2004-06~');
      expect(e6_2.toString(), '2004-06~');
      final e6_3 = Edtf.parse('2004-06-11%');
      expect(e6_3.toString(), '2004-06-11%');
      final e7_1 = Edtf.parse('201X');
      expect(e7_1.toString(), '201X');
      final e7_2 = Edtf.parse('20XX');
      expect(e7_2.toString(), '20XX');
      final e7_3 = Edtf.parse('2004-XX');
      expect(e7_3.toString(), '2004-XX');
      final e7_4 = Edtf.parse('1985-04-XX');
      expect(e7_4.toString(), '1985-04-XX');
      final e7_5 = Edtf.parse('1985-XX-XX');
      expect(e7_5.toString(), '1985-XX-XX');
      final e8_1 = Edtf.parse('1985-04-12/..');
      expect(e8_1.toString(), '1985-04-12/..');
      final e8_2 = Edtf.parse('1985-04/..');
      expect(e8_2.toString(), '1985-04/..');
      final e8_3 = Edtf.parse('1985/..');
      expect(e8_3.toString(), '1985/..');
      final e8_4 = Edtf.parse('../1985-04-12');
      expect(e8_4.toString(), '../1985-04-12');
      final e8_5 = Edtf.parse('../1985-04');
      expect(e8_5.toString(), '../1985-04');
      final e8_6 = Edtf.parse('../1985');
      expect(e8_6.toString(), '../1985');
      final e8_7 = Edtf.parse('1985-04-12/');
      expect(e8_7.toString(), '1985-04-12/');
      final e8_8 = Edtf.parse('1985-04/');
      expect(e8_8.toString(), '1985-04/');
      final e8_9 = Edtf.parse('1985/');
      expect(e8_9.toString(), '1985/');
      final e8_10 = Edtf.parse('/1985-04-12');
      expect(e8_10.toString(), '/1985-04-12');
      final e8_11 = Edtf.parse('/1985-04');
      expect(e8_11.toString(), '/1985-04');
      final e8_12 = Edtf.parse('/1985');
      expect(e8_12.toString(), '/1985');
      final e9_1 = Edtf.parse('-1985');
      expect(e9_1.toString(), '-1985');
      final e10_1 = Edtf.parse('Y-17E7');
      expect(e10_1.toString(), 'Y-17E7');
      final e11_1 = Edtf.parse('1950S2');
      expect(e11_1.toString(), '1950S2');
      final e11_2 = Edtf.parse('Y171010000S3');
      expect(e11_2.toString(), 'Y171010000S3');
      final e11_3 = Edtf.parse('Y3388E2S3');
      expect(e11_3.toString(), 'Y3388E2S3');
      final e12_1 = Edtf.parse('2001-34');
      expect(e12_1.toString(), '2001-34');
      final e13_1 = Edtf.parse('[1667,1668,1670..1672]');
      expect(e13_1.toString(), '[1667,1668,1670..1672]');
      final e13_2 = Edtf.parse('[..1760-12-03]');
      expect(e13_2.toString(), '[..1760-12-03]');
      final e13_3 = Edtf.parse('[1760-12..]');
      expect(e13_3.toString(), '[1760-12..]');
      final e13_4 = Edtf.parse('[1760-01,1760-02,1760-12..]');
      expect(e13_4.toString(), '[1760-01,1760-02,1760-12..]');
      final e13_5 = Edtf.parse('[1667,1760-12]');
      expect(e13_5.toString(), '[1667,1760-12]');
      final e13_6 = Edtf.parse('[..1984]');
      expect(e13_6.toString(), '[..1984]');
      final e13_7 = Edtf.parse('{1667,1668,1670..1672}');
      expect(e13_7.toString(), '{1667,1668,1670..1672}');
      final e13_8 = Edtf.parse('{1960,1961-12}');
      expect(e13_8.toString(), '{1960,1961-12}');
      final e13_9 = Edtf.parse('{..1984}');
      expect(e13_9.toString(), '{..1984}');
      final e14_1 = Edtf.parse('2004-06-11%');
      expect(e14_1.toString(), '2004-06-11%');
      final e14_2 = Edtf.parse('2004-06~-11');
      expect(e14_2.toString(), '2004-06~-11');
      final e14_3 = Edtf.parse('2004?-06-11');
      expect(e14_3.toString(), '2004?-06-11');
      final e14_4 = Edtf.parse('?2004-06-~11');
      expect(e14_4.toString(), '?2004-06-~11');
      final e14_5 = Edtf.parse('2004-%06-11');
      expect(e14_5.toString(), '2004-%06-11');
      final e15_1 = Edtf.parse('156X-12-25');
      expect(e15_1.toString(), '156X-12-25');
      final e15_2 = Edtf.parse('15XX-12-25');
      expect(e15_2.toString(), '15XX-12-25');
      final e15_3 = Edtf.parse('XXXX-12-XX');
      expect(e15_3.toString(), 'XXXX-12-XX');
      final e15_4 = Edtf.parse('1XXX-XX');
      expect(e15_4.toString(), '1XXX-XX');
      final e15_5 = Edtf.parse('1XXX-12');
      expect(e15_5.toString(), '1XXX-12');
      final e15_6 = Edtf.parse('1984-1X');
      expect(e15_6.toString(), '1984-1X');
      final e16_1 = Edtf.parse('2004-06-~01/2004-06-~20');
      expect(e16_1.toString(), '2004-06-~01/2004-06-~20');
      final e16_2 = Edtf.parse('2004-06-XX/2004-07-03');
      expect(e16_2.toString(), '2004-06-XX/2004-07-03');
    });
  });
}


lib/src/edtf_base.dart
import 'dart:math';
import 'package:meta/meta.dart';

/// Abstract class for handling any kind of Extended Date/Time Format.
abstract class Edtf {
  Edtf._();

  /// Parses [s] to a subtype of Edtf based on the string
  factory Edtf.parse(String s) {
    if (RegExp('^\\[.*\\]\$').hasMatch(s)) {
      // OneOf
      return EdtfOneOf.parse(s);
    } else if (RegExp('^{.*}\$').hasMatch(s)) {
      // Every
      return EdtfEvery.parse(s);
    } else if (RegExp('/').hasMatch(s)) {
      // Interval
      return EdtfInterval.parse(s);
    } else {
      // Date
      return EdtfDate.parse(s);
    }
  }

  /// Inner: in enclosed dates, interval rules are different
  factory Edtf._parseInner(String s) {
    if (RegExp('\\.\\.').hasMatch(s)) {
      // Interval
      return EdtfInterval._parseInner(s);
    } else {
      // Date
      return EdtfDate.parse(s);
    }
  }

  @override
  String toString() {
    return super.toString();
  }

  String _toInnerString() {
    return toString();
  }
}

/// An abstract class for handling cases where the date can be from a set
abstract class EdtfSet extends Edtf {
  /// The set of dates which can be used in the representation.
  final List<Edtf> values;
  EdtfSet._([this.values]) : super._();

  static List<Edtf> _parseDates(String s) {
    final dates = <Edtf>[];
    for (final elem in s.split(',')) {
      dates.add(Edtf._parseInner(elem));
    }
    return dates;
  }

  @override
  String toString() {
    var ret = '';
    for (final edtf in values) {
      ret += edtf._toInnerString() + ',';
    }
    ret = ret.substring(0, ret.length - 1);
    return ret;
  }
}

/// A class for handling cases where the edtf string represents all of the dates
/// from a set.
class EdtfEvery extends EdtfSet {
  /// Constructor for directly creating an object with the given values
  EdtfEvery._(List<Edtf> values) : super._(values);
  static final _regExp = RegExp('^{(.*)}\$');

  /// Parse a '{...}' like string to an EdtfEvery object
  factory EdtfEvery.parse(String s) {
    return EdtfEvery._(EdtfSet._parseDates(_regExp.firstMatch(s).group(1)));
  }

  @override
  String toString() {
    return '{' + super.toString() + '}';
  }
}

/// A class for handling cases where the edtf string represents one of the dates
/// from a set
class EdtfOneOf extends EdtfSet {
  /// Constructor for directly creating an object with the given values
  EdtfOneOf._(List<Edtf> values) : super._(values);

  /// Parses a '[...]' like string to an EdtfOneOf object
  factory EdtfOneOf.parse(String s) {
    final _regExp = RegExp('^\\[(.*)\\]\$');
    return EdtfOneOf._(EdtfSet._parseDates(_regExp.firstMatch(s).group(1)));
  }

  @override
  String toString() {
    return '[' + super.toString() + ']';
  }
}

/// A class for handling intervals as specified by the edtf standard.
class EdtfInterval extends Edtf {
  /// Start of the interval. Can be null, if it's unknown or open
  final EdtfDate start;

  /// End of the interval. Can be null, if it's unknown or open
  final EdtfDate end;

  /// True if the start of the interval is open,
  /// false if it has a start or the start is unknown
  final bool openStart;

  /// True if the end of the interval is open,
  /// false if it has an end or the end is unknown
  final bool openEnd;

  /// Creates an object with the given parameters
  EdtfInterval._(this.start, this.openStart, this.end, this.openEnd)
      : super._();

  /// Parses a '?/?' like string to an interval
  factory EdtfInterval.parse(String s) {
    final dates = s.split('/');
    EdtfDate start, end;
    bool openStart, openEnd;
    assert(dates.length == 2);
    if (dates[0] == '') {
      start = null;
      openStart = false;
    } else if (dates[0] == '..') {
      start = null;
      openStart = true;
    } else {
      start = EdtfDate.parse(dates[0]);
      openStart = false;
    }
    if (dates[1] == '') {
      end = null;
      openEnd = false;
    } else if (dates[1] == '..') {
      end = null;
      openEnd = true;
    } else {
      end = EdtfDate.parse(dates[1]);
      openEnd = false;
    }
    return EdtfInterval._(start, openStart, end, openEnd);
  }

  /// Parses a '?..?' like string to an interval
  factory EdtfInterval._parseInner(String s) {
    final dates = s.split('..');
    EdtfDate start, end;
    bool openStart, openEnd;
    assert(dates.length == 2);
    if (dates[0] == '') {
      start = null;
      openStart = true;
    } else {
      start = EdtfDate.parse(dates[0]);
      openStart = false;
    }
    if (dates[1] == '') {
      end = null;
      openEnd = true;
    } else {
      end = EdtfDate.parse(dates[1]);
      openEnd = false;
    }
    return EdtfInterval._(start, openStart, end, openEnd);
  }

  @override
  String toString() {
    var ret = '';
    if (openStart) {
      ret += '..';
    }
    if (start != null) {
      ret += start.toString();
    }
    ret += '/';
    if (end != null) {
      ret += end.toString();
    }
    if (openEnd) {
      ret += '..';
    }
    return ret;
  }

  @override
  String _toInnerString() {
    var ret = '';
    if (start != null) {
      ret += start.toString();
    }
    ret += '..';
    if (end != null) {
      ret += end.toString();
    }
    return ret;
  }
}

/// A class for handling dates according to the edtf specification.
class EdtfDate extends Edtf {
  /// The year part of the date with the modifiers
  final EdtfNumber year;

  /// The month part of the date with the modifiers. Can be null, if not given
  final EdtfNumber month;

  /// The day part of the date with the modifiers. Can be null, if not given
  final EdtfNumber day;

  /// The time part of the date. Can be null, if not given
  final EdtfTime time;

  /// Constructs a EdtfDate objects from the given parameters
  EdtfDate._(this.year, [this.month, this.day, this.time]) : super._();

  /// Parses a date according to edtf specification.
  factory EdtfDate.parse(String s) {
    EdtfTime etime;
    if (s.contains('T')) {
      final time = s.substring(s.indexOf('T') + 1);
      etime = EdtfTime.parse(time);
      s = s.substring(0, s.indexOf('T'));
    } else {
      etime = null;
    }

    final numberLike = RegExp('[\\dXu]+');
    final parts = s.split('-'); // What if -1000-10-10?
    if (!parts[0].contains(numberLike)) {
      final year = parts[0] + '-' + parts[1];
      parts.removeAt(0);
      parts[0] = year;
    }
    EdtfNumber year, month, day;
    year = EdtfNumber.parse(parts[0]);
    if (parts.length >= 2) {
      month = EdtfNumber.parse(parts[1]);
      year = EdtfNumber._addGroup(year, month);
    }
    if (parts.length >= 3) {
      day = EdtfNumber.parse(parts[2]);
      month = EdtfNumber._addGroup(month, day);
      year = EdtfNumber._addGroup(year, day);
    }
    return EdtfDate._(year, month, day, etime);
  }

  @override
  String toString() {
    var ret = year.toString();
    if (month != null) {
      ret += '-' + month.toString();
    }
    if (day != null) {
      ret += '-' + day.toString();
    }
    if (time != null) {
      ret += 'T' + time.toString();
    }
    return ret;
  }
}

/// A class for handling time according to the edtf specification
class EdtfTime {
  /// The hour part of the time
  final int hour;

  /// The minute part of the time
  final int minutes;

  /// The seconds part of the time
  final int seconds;
  final int _shiftHour;
  final int _shiftMinute;

  /// The hour part of the time shift. Can be null, if time is local
  int get shiftHour {
    switch (shiftLevel) {
      case shiftLevelUTC:
        return 0;
      case shiftLevelHour:
      case shiftLevelMinute:
        return _shiftHour;
      default: // shiftLevel == shiftLevelLocal or error
        return null;
    }
  }

  /// The minute part of the time shift. Can be null, if time is local
  int get shiftMinute {
    switch (shiftLevel) {
      case shiftLevelUTC:
      case shiftLevelHour:
        return 0;
      case shiftLevelMinute:
        return _shiftMinute;
      default: // shiftLevel == shiftLevelLocal or error
        return null;
    }
  }

  /// The precision of the shift where the time is stored at. Can be
  /// Local, UTC, Hour or Minute precision with the EdtfTime.shiftLevel* value.
  final int shiftLevel;

  /// Constant for representing Local shift value
  static const shiftLevelLocal = 0;

  /// Constant for representing UTC time ('Z' at the end)
  static const shiftLevelUTC = 1;

  /// Constanf for representing hour precision shifted time
  static const shiftLevelHour = 2;

  /// Constanf for representing minute precision shifted time
  static const shiftLevelMinute = 3;

  /// Constructor for creating an object from the given values
  EdtfTime._(this.hour, this.minutes, this.seconds, this.shiftLevel,
      [this._shiftHour, this._shiftMinute]);

  /// Creates an EdtfTime object from a 'XX:XX:XX' + shift? like string
  factory EdtfTime.parse(String time) {
    final parts = time.split(':');
    final hour = int.parse(parts[0]);
    final minutes = int.parse(parts[1]);
    final seconds = int.parse(parts[2].substring(0, 2));
    if (parts[2].length <= 2) {
      // Local time
      final shiftLevel = EdtfTime.shiftLevelLocal;
      return EdtfTime._(hour, minutes, seconds, shiftLevel);
    } else if (parts[2][2] == 'Z') {
      // UTC time
      final shiftLevel = EdtfTime.shiftLevelUTC;
      return EdtfTime._(hour, minutes, seconds, shiftLevel);
    } else {
      // Shift is added
      final shiftHour = int.parse(parts[2].substring(2));
      if (parts.length <= 3) {
        // Has shiftHour precision
        final shiftLevel = EdtfTime.shiftLevelHour;
        return EdtfTime._(hour, minutes, seconds, shiftLevel, shiftHour);
      } else {
        // Has shiftHour:shiftMinute format
        final shiftMinute = int.parse(parts[3]);
        final shiftLevel = EdtfTime.shiftLevelMinute;
        return EdtfTime._(
            hour, minutes, seconds, shiftLevel, shiftHour, shiftMinute);
      }
    }
  }

  @override
  String toString() {
    var ret = '';
    ret += _toFixedString(hour, 2);
    ret += ':';
    ret += _toFixedString(minutes, 2);
    ret += ':';
    ret += _toFixedString(seconds, 2);
    if (shiftLevel == shiftLevelLocal) {
      return ret;
    } else if (shiftLevel == shiftLevelUTC) {
      return ret + 'Z';
    } else if (shiftLevel == shiftLevelHour) {
      ret += _toFixedString(_shiftHour, 2, plusSign: true);
      return ret;
    } else if (shiftLevel == shiftLevelMinute) {
      ret += _toFixedString(_shiftHour, 2, plusSign: true);
      ret += ':';
      ret += _toFixedString(_shiftMinute, 2);
      return ret;
    } else {
      throw ArgumentError('Unexpected shiftLevel');
    }
  }
}

/// Converts a nuber to a string with leading zeroes
String _toFixedString(int number, int width, {bool plusSign = false}) {
  var ret = '';
  if (plusSign && number >= 0) {
    ret += '+';
  } else if (number < 0) {
    ret += '-';
  }
  ret += number.abs().toString().padLeft(width, '0');
  return ret;
}

/// A class for handling the number-like parts of an edtf date
class EdtfNumber {
  /// The value of the number from string,
  /// before shifting according to exp, and with '0' for unknown parts.
  final int value;

  /// The exponent of the value, from the 'E' number. Can be null, if not given
  final int exp;

  /// The significant digits in the value, from the 'S' number.
  /// Can be null, if not given
  final int significance;

  /// The mask of the unknown characters in the value, with '.' in place of
  /// 'X' (or 'u')
  final String unspecMask;

  /// True if the number has '~' or '%', so it's approximate.
  final bool localApprox;

  /// True if the number has '?' or '%' characters, stating it is uncertain
  final bool localUncert;

  /// True if the number is approximate because of a group qualifier
  final bool groupApprox;

  /// True if the number is uncertain because of a group qualifier
  final bool groupUncert;

  /// The real value of the number, shifted according to the exponent
  int get realValue => value * pow(10, exp).toInt();

  /// True if the number is an approximation
  bool get approx => localApprox | groupApprox;

  /// True if the number is uncertain
  bool get uncert => localUncert | groupUncert;

  // TODO replace like modifier constructor
  EdtfNumber._(
      {@required this.value,
      this.exp,
      this.significance,
      this.unspecMask,
      this.localApprox,
      this.localUncert,
      this.groupApprox,
      this.groupUncert});

  factory EdtfNumber._addGroup(EdtfNumber target, EdtfNumber prev) {
    return EdtfNumber._(
        value: target.value,
        exp: target.exp,
        significance: target.significance,
        unspecMask: target.unspecMask,
        localApprox: target.localApprox,
        localUncert: target.localUncert,
        groupApprox: target.groupApprox | prev.groupApprox,
        groupUncert: target.groupUncert | prev.groupUncert);
  }

  /// Parses the number from a string
  factory EdtfNumber.parse(String s) {
    // Group 1: value with missing digits
    // Group 3: exp
    // Group 5: prec
    final nums = RegExp(r'(-?[\dXu]+)([Ee](\d+))?([Sp](\d+))?').firstMatch(s);
    final value = int.parse(nums.group(1).replaceAll(RegExp('[Xu]'), '0'));
    final exp = nums.group(3) != null ? int.parse(nums.group(3)) : null;
    final precision = nums.group(5) != null ? int.parse(nums.group(5)) : null;
    final unspecMask = nums.group(1).replaceAll(RegExp('[Xu]'), '.');

    final localApprox = s.contains('~') || s.contains('%');
    final localUncert = s.contains('?') || s.contains('%');
    final groupApprox = s.endsWith('~') || s.endsWith('%');
    final groupUncert = s.endsWith('?') || s.endsWith('%');

    return EdtfNumber._(
        value: value,
        exp: exp,
        significance: precision,
        unspecMask: unspecMask,
        localApprox: localApprox,
        localUncert: localUncert,
        groupApprox: groupApprox,
        groupUncert: groupUncert);
  }
  @override
  String toString() {
    var ret = '';
    if (value < -9999 || 9999 < value || exp != null) ret += 'Y';
    ret += unspecMask.replaceAll('.', 'X');
    if (exp != null) ret += 'E' + exp.toString();
    if (significance != null) ret += 'S' + significance.toString();
    final groupChar = groupApprox ? groupUncert ? '%' : '~' : '?';
    if ((localApprox && groupApprox) || (localUncert && groupUncert)) {
      ret += groupChar;
    }
    final localChar = localApprox ? localUncert ? '%' : '~' : '?';
    if ((localApprox && !groupApprox) || (localUncert && !groupUncert)) {
      ret = localChar + ret;
    }
    return ret;
  }
}

/// The special values for months, according to the specification
const edtfMonthNames = {
  21: 'Spring',
  22: 'Summer',
  23: 'Autumn',
  24: 'Winter',
  25: 'Spring - Northern Hemisphere',
  26: 'Summer - Northern Hemisphere',
  27: 'Autumn - Northern Hemisphere',
  28: 'Winter - Northern Hemisphere',
  29: 'Spring - Southern Hemisphere',
  30: 'Summer - Southern Hemisphere',
  31: 'Autumn - Southern Hemisphere',
  32: 'Winter - Southern Hemisphere',
  33: 'Quarter 1 (3 months in duration)',
  34: 'Quarter 2 (3 months in duration)',
  35: 'Quarter 3 (3 months in duration)',
  36: 'Quarter 4 (3 months in duration)',
  37: 'Quadrimester 1 (4 months in duration)',
  38: 'Quadrimester 2 (4 months in duration)',
  39: 'Quadrimester 3 (4 months in duration)',
  40: 'Semestral 1 (6 months in duration)',
  41: 'Semestral 2 (6 months in duration)',
};


example/edtf_example.dart
import 'package:edtf/edtf.dart';

main() {
  /// Parse a date in the EDTF format
  final date = EdtfDate.parse('1985-05%-13?');

  /// Check that the year is an approximation.
  if (date.year.approx) {
    /// If it is an approximation, then print it
    print(date.toString());
  }

  /// Parse an EDTF formatted interval
  final interval = EdtfInterval.parse('../2012');

  /// If the interval is open started and the end is a known date...
  if (interval.openStart || interval.end != null) {
    /// ... print the end date.
    print(interval.end.toString());
  }

  /// Parse a set of EDTF dates/intervals, where the real date is one of them.
  final oneOf = EdtfOneOf.parse('[1981, 1982-~04, 1983-1987]');

  /// Iterate through the dates/intervals
  for (final elem in oneOf.values) {
    /// If the elem is a date print it in this format
    if (elem is EdtfDate) {
      print('Date: ' + elem.toString());
    }

    /// If the elem is an interval just print the start date
    if (elem is EdtfInterval) {
      print('Interval start date: ' +
          (elem.openStart ? 'open' : elem.start.toString()));
      // An interval either has an known start date or open started
    }
  }
}
